net/sched/sch_gkprioq.c

```diff
@@ -0,0 +1,395 @@
+ /*
+ * net/sched/sch_gkprioq.c  Gatekeeper Priority Queue.
+ *
+ *		This program is free software; you can redistribute it and/or
+ *		modify it under the terms of the GNU General Public License
+ *		as published by the Free Software Foundation; either version
+ *		2 of the License, or (at your option) any later version.
+ *
+ * Authors:	Nishanth Devarajan, <ndev_2021@gmail.com>
+ *	        original idea by Michel Machado, Cody Doucette and Meng Xiang
+ *
+ */
+
+#include <linux/string.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/types.h>
+#include <linux/kernel.h>
+#include <linux/errno.h>
+#include <linux/skbuff.h>
+#include <net/pkt_sched.h>
+#include <net/sch_generic.h>
+#include <net/dsfield.h>
+#include <net/inet_ecn.h>
+#include <asm/byteorder.h>
+
+/* Priority markers */
+#define MAX_PRIORITY 63
+#define HALF_MAX_PRIORITY 31
+
+
+
+/*	  Gatekeeper Priority Queue
+ *	=================================
+ *
+ *
+ * Gatekeeper is an open source DoS defense software which sends traffic based   * on an assigned priority for the client's requeust. Assigning priorities to    * packets enables us to give low priority to clients who quickly send many      * requests, so  that we can service legitimate clients (who typically send      * requests more slowly). The DoS scenario is when assigning low priorities      * helps block out malicious senders, who could otherwise overwhelm Gatekeeper   * with a very high number of requests, denying service to the requests of 
+ * legitimate users.
```

Comment:
You should make a round of changes trying to address the comments here and then update it all at once.